include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h

```diff
@@ -225,17 +226,45 @@ __get_accessor_size(const _Accessor& __accessor)
 #endif
 }
 
-template <typename _Item>
-constexpr void
-__group_barrier(_Item __item)
+// TODO: switch to SYCL 2020 with DPC++ compiler.
+// SYCL 1.2.1 version is used due to having an API with a local memory fence,
+// which gives better performance on Intel GPUs.
+// The performance gap is negligible since
+// https://github.com/intel/intel-graphics-compiler/commit/ed639f68d142bc963a7b626badc207a42fb281cb (Aug 20, 2024)
+// But the fix is not a part of the LTS GPU drivers (Linux) yet.
+#if !defined(ONEDPL_SYCL121_GROUP_BARRIER)
+#    if _ONEDPL_LIBSYCL_VERSION
+#        define ONEDPL_SYCL121_GROUP_BARRIER 1
+#    else
+#        define ONEDPL_SYCL121_GROUP_BARRIER 0
```

Comment:
This is the only think I see as "risky" in the PR now that we are past code freeze.  

This changes the default behavior for non-intel LLVM compilers to assume sycl 2020 group barrier support, where previous to this PR, we were assuming sycl 1.2.1 support.

Would it make sense to have a `SYCL_LANGUAGE_VERSION >= 202000` check here and decide for non-intel compilers that way?

(This remains a change which may not be "required" at a time where we should resist making changes).

Reply:
I think we have the following decision to make for the else branch here:
1) Minimize change post code freeze
`#define ONEDPL_SYCL121_GROUP_BARRIER 1`

2) Assume sycl 2020 for non intel llvm compiler 
`#define ONEDPL_SYCL121_GROUP_BARRIER 0`

3) Check sycl spec version macro and decide based on that
`#define ONEDPL_SYCL121_GROUP_BARRIER SYCL_LANGUAGE_VERSION >= 202000`


I don't have as good a grasp on what should be expected for non intel-llvm sycl compilers, so I'm interested what others think here.

Reply:
I suggest making a decision on Monday as it is late. In the meantime, I will test the PR using icx and acpp.

My personal preference is option 2: oneDPL declares SYCL 2020 compatibility, so this change can be treated as a bug fix, while there is a backup option to manually set `ONEDPL_SYCL121_GROUP_BARRIER` for those who rely on these barriers.

Reply:
Does acpp correctly define `SYCL_LANGUAGE_VERSION`?

If so, would (3) be a more sycl specification compliant version of basically the same as (2)?

Reply:
> Does acpp correctly define SYCL_LANGUAGE_VERSION?

It does: https://github.com/AdaptiveCpp/AdaptiveCpp/blob/e1c6c6ede4c17877a2005f876afc869f11fba5be/include/hipSYCL/sycl/sycl.hpp#L27C1-L27C38

> If so, would (3) be a more sycl specification compliant version of basically the same as (2) at least for acpp?

Agree (3) looks better than (2) with this additional check. The check should be slightly different, though. If it is SYCL 2020, then define `ONEDPL_SYCL121_GROUP_BARRIER` to `0`.

Reply:
yes you are correct.  Thanks for the correction.

Reply:
I checked that ACPP supports SYCL 2020 barriers. I am now slightly inclined towards keeping option (2) as we want to have only SYCL 2020 barriers in the future, and we do not now other implementations still having only SYCL 1.2.1 barriers, but which are SYCL 2020 compliant in other places.

Reply:
I'm OK with that option, but lets leave a comment explaining it.

Reply:
Added:
> // For safety, assume that other SYCL implementations comply with SYCL 2020, which is a oneDPL requirement.